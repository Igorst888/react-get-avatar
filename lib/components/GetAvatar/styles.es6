export default {
  wrapper: {
    position: 'relative',
  },

  canvas: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: 0,
    height: 0,
    overflow: 'hidden',
  },

  canvasActive: {
    width: '100%',
    height: '100%',
    backgroundColor: 'white',
  },

  button: {
    display: 'inline-block',
    verticalAlign: 'top',
    width: '50%',
    height: '50px',
    backgroundColor: 'steelblue',
    fontSize: '16px',
    lineHeight: '50px',
    color: 'white',
    textAlign: 'center',
    cursor: 'pointer',
  },

  buttonOnly: {
    width: '100%',
  },

  buttonFirst: {
    width: 'calc(50% - 1px)',
    marginRight: '1px',
  },

  buttonUpload: {
    position: 'relative',
    overflow: 'hidden',
  },

  inputUpload: {
    position: 'absolute',
    width: '0.1px',
    height: '0.1px',
    opacity: '0',
    overflow: 'hidden',
    zIndex: '-1',
  },

  capture: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    opacity: '0.5',
    backgroundImage: " url(\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='200px' height='200px'%3E%3Cpolygon points='5,5 195,5 195,195 5,195' fill='none' stroke-width='10' stroke='%23FFF' stroke-linejoin='round' stroke-dasharray='100 90' stroke-dashoffset='50' /%3E%3C/svg%3E\")",
    backgroundPosition: '50% 50%',
    backgroundSize: '200px 200px',
    backgroundRepeat: 'no-repeat',
    cursor: 'pointer',
  },


};
