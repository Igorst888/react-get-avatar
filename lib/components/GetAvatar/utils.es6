export default function getDataURL({ data, width, height }) {
  const canvas = document.createElement('canvas');
  canvas.width = width;
  canvas.height = height;

  const ctx = canvas.getContext('2d');
  ctx.putImageData(data, 0, 0);

  return canvas.toDataURL('image/jpeg', 0.5);
}
