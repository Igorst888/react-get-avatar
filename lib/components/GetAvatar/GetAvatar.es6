import React from 'react';
import PropTypes from 'prop-types';

import GetAvatarWebcam from '../GetAvatarWebcam/GetAvatarWebcam';
import GetAvatarCanvas from '../GetAvatarCanvas/GetAvatarCanvas';
import styles from './styles';

export default class GetAvatar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      webcamReady: false,
      webcamStarted: false,
      webcamElement: null,
      canvasReady: false,
      canvasActive: false,
      canvasElement: null,
      canvasSource: null,
    };
  }

  componentWillMount() {
    this.listeners = {
      onClickWebcamToggle: this.onClickWebcamToggle.bind(this),
      onClickWebcamCapture: this.onClickWebcamCapture.bind(this),
      onChangeImageUpload: this.onChangeImageUpload.bind(this),
      onChangeCanvasImage: this.onChangeCanvasImage.bind(this),
      onClickSave: this.onClickSave.bind(this),
      onClickCancel: this.onClickCancel.bind(this),
    };
  }

  onWebcamReady(elem) {
    this.setState({
      webcamReady: true,
      webcamElement: elem,
    });
  }

  onCanvasReady(elem) {
    this.setState({
      canvasReady: true,
      canvasElement: elem,
    });
  }

  onClickWebcamToggle() {
    const { webcamReady, webcamStarted } = this.state;

    if (!webcamReady) return;
    this.setState({
      webcamStarted: !webcamStarted,
    });
  }

  onClickWebcamCapture() {
    const { webcamElement } = this.state;

    this.setState({
      canvasSource: webcamElement,
    });
  }

  onChangeImageUpload(event) {
    const reader = new FileReader();

    reader.onload = (readerEvent) => {
      const image = new Image();
      image.onload = () => {
        this.setState({
          canvasSource: image,
        });
      };
      image.src = readerEvent.target.result;
    };

    reader.readAsDataURL(event.target.files[0]);
  }

  onChangeCanvasImage() {
    this.setState({
      canvasActive: true,
    });
  }

  onClickSave() {
    const { handleGet, imageType, imageQuality } = this.props;

    const dataURI = this.state.canvasElement.toDataURL(imageType, imageQuality);

    if (handleGet) handleGet(dataURI);
  }

  onClickCancel() {
    this.setState({
      canvasActive: false,
      canvasSource: null,
    });
  }

  wrapperStyles() {
    const { width, height } = this.props;
    return {
      ...styles.wrapper,
      width: `${width}px`,
      height: `${height}px`,
    };
  }

  canvasStyles() {
    const { canvasActive } = this.state;

    return canvasActive ? { ...styles.canvas, ...styles.canvasActive } : styles.canvas;
  }

  rootStyles() {
    const { className } = this.props;
    return className ? { className } : null;
  }

  navStyles() {
    const className = this.props.navClassName;
    return className ? { className } : null;
  }

  toggleStyles() {
    const className = this.props.buttonClassName.toggle;
    if (className) return { className };

    const style = { ...styles.button, ...styles.buttonFirst };
    return { style };
  }

  uploadStyles() {
    const className = this.props.buttonClassName.upload;
    if (className) return { className };

    const { webcamReady, canvasReady } = this.state;
    if (webcamReady && canvasReady) {
      const style = { ...styles.button, ...styles.buttonUpload };
      return { style };
    }
    const style = { ...styles.button, ...styles.buttonOnly, ...styles.buttonUpload };
    return { style };
  }

  saveStyles() {
    const className = this.props.buttonClassName.save;
    if (className) return { className };

    const style = { ...styles.button, ...styles.buttonFirst };
    return { style };
  }

  cancelStyles() {
    const className = this.props.buttonClassName.cancel;
    if (className) return { className };

    const style = styles.button;
    return { style };
  }

  render() {
    const { width, height } = this.props;
    const {
      webcamReady, webcamStarted,
      canvasReady, canvasSource, canvasActive,
    } = this.state;

    return (
      <div { ...this.rootStyles() }>
        <div style={ this.wrapperStyles() }>
          <GetAvatarWebcam
            handleReady={ (elem) => { this.onWebcamReady(elem); } }
            isActive={ webcamStarted && !canvasActive }
          />
          { webcamReady && webcamStarted && canvasReady && <div
            style={ styles.capture }
            onClick={ this.listeners.onClickWebcamCapture }
          /> }
          <div style={ this.canvasStyles() }>
            <GetAvatarCanvas
              source={ canvasSource }
              width={ width }
              height={ height }
              handleReady={ (elem) => { this.onCanvasReady(elem); } }
              handleDraw={ this.listeners.onChangeCanvasImage }
            />
          </div>
        </div>
        { !canvasActive && <div { ...this.navStyles() }>
          { webcamReady && canvasReady && <div
            { ...this.toggleStyles() }
            onClick={ this.listeners.onClickWebcamToggle }
          >
            { webcamStarted ? 'Stop Webcam' : 'Start Webcam' }
          </div> }
          <label { ...this.uploadStyles() }>
            { 'Upload File' }
            <input
              style={ styles.inputUpload }
              type='file'
              onChange={ this.listeners.onChangeImageUpload }
            />
          </label>
        </div> }
        { canvasActive && <div { ...this.navStyles() }>
          <div
            { ...this.saveStyles() }
            onClick={ this.listeners.onClickSave }
          >
            { 'Save' }
          </div>
          <div
            { ...this.cancelStyles() }
            onClick={ this.listeners.onClickCancel }
          >
            { 'Cancel' }
          </div>
        </div> }
      </div>
    );
  }
}

const { func, oneOf, number, shape, string } = PropTypes;

GetAvatar.propTypes = {
  width: number.isRequired,
  height: number.isRequired,
  imageType: oneOf(['image/jpeg', 'image/png', 'image/gif', 'image/webp']),
  imageQuality: number,
  className: string,
  navClassName: string,
  buttonClassName: shape({
    toggle: string,
    upload: string,
    save: string,
    cancel: string,
  }),
  handleGet: func.isRequired,
};

GetAvatar.defaultProps = {
  width: 0,
  height: 0,
  imageType: 'image/jpeg',
  imageQuality: 0.5,
  buttonClassName: {},
  handleGet: () => {},
};
