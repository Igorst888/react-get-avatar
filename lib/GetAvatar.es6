import React from 'react';
import PropTypes from 'prop-types';

import { getUserMedia, getMediaSources, createScreenshot } from './utils';
import styles from './styles';

export default class GetAvatar extends React.Component {
  constructor(props) {
    super(props);

    navigator.getUserMedia = getUserMedia();

    this.state = {
      started: false,
      stream: null,
      sources: null,
      sourceIndex: 0,
      dataURI: null,
    };
  }

  componentWillMount() {
    this.listeners = {
      handleVideo: this.handleVideo.bind(this),
      handleError: this.handleError.bind(this),
      onClickVideo: this.onClickVideo.bind(this),
      onClickFlip: this.onClickFlip.bind(this),
    };
  }

  componentWillUnmount() {
    this.stop();
  }

  componentDidMount() {
    getMediaSources().then((sources) => {
      this.setState({ sources });
      this.start();
    }).catch(() => {

    });
  }

  onClickVideo() {
    const { started } = this.state;

    if (!started) return;

    const { handleGet } = this.props;

    const dataURI = createScreenshot(this.video);
    const processed = true;

    this.setState({ dataURI, processed });

    setTimeout(() => {
      this.setState({
        dataURI: null,
        processed: false,
      });
    }, 250);

    if (handleGet) handleGet(dataURI);
  }

  onClickFlip() {
    const { sources } = this.state;
    const sourcesCount = sources.length;

    if (sourcesCount < 2) return;

    let { sourceIndex } = this.state;

    sourceIndex += 1;

    if (sourceIndex > sourcesCount - 1) {
      sourceIndex = 0;
    }

    this.setState({ sourceIndex });

    this.stop();
    this.start();
  }

  getRootStyles() {
    const { width, height } = this.props;

    const rootStyles = styles.root;

    if (!width && !height) {
      rootStyles.paddingTop = '75%';

      return rootStyles;
    }

    if (width) {
      rootStyles.width = `${width}px`;
    }

    if (height) {
      rootStyles.height = `${height}px`;
    } else {
      rootStyles.height = '100%';
    }

    return rootStyles;
  }

  getScreenStyles() {
    const { processed } = this.state;

    if (!processed) {
      return styles.screen;
    }

    return {
      ...styles.screen,
      ...styles.processed,
    };
  }

  handleVideo(stream) {
    this.setState({ stream });
  }

  handleError(err) {
    console.error(`getUserMedia error (${err.name}: ${err.message})`);
  }

  start() {
    const { sources, sourceIndex } = this.state;
    const constraints = { video: true };

    if (sources && sources.length > 1) {
      constraints.video = {
        optional: [{ sourceId: sources[sourceIndex].deviceId }],
      };
    }

    navigator.getUserMedia(constraints, this.listeners.handleVideo, this.listeners.handleError);

    this.setState({ started: true });
  }

  stop() {
    const { stream } = this.state;

    if (!stream) return;

    const track = stream.getTracks()[0];

    if (track) track.stop();

    this.setState({ started: false });
  }

  render() {
    if (!navigator.getUserMedia) {
      return (
        <div style={ this.getRootStyles() }>
          <span style={{ color: 'white' }}>Webcam not supported</span>
        </div>
      );
    }

    const { stream, sources, dataURI } = this.state;
    const hasScreenshot = dataURI !== null;

    const videoSrc = stream ? window.URL.createObjectURL(stream) : null;
    const hasVideo = videoSrc !== null;

    const hasSources = hasVideo && (sources && sources.length > 1);

    return (
      <div style={ this.getRootStyles() }>
        { hasVideo && <video
          ref={ (c) => { this.video = c; }}
          src={ videoSrc }
          style={ styles.video }
          autoPlay
        /> }
        { hasScreenshot && <img
          style={ styles.screenshot }
          src={ dataURI }
          alt=''
        /> }
        { hasVideo && <div
          style={ this.getScreenStyles() }
          onClick={ this.listeners.onClickVideo }
        /> }
        { hasSources && <div
          style={ styles.flip }
          onClick={ this.listeners.onClickFlip }
        /> }
      </div>
    );
  }
}

GetAvatar.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  handleGet: PropTypes.func,
};

GetAvatar.defaultProps = {
  width: 0,
  height: 0,
  handleGet: () => {},
};
