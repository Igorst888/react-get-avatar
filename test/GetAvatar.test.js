/* eslint-env jest */

import { shallow } from 'enzyme';

import React from 'react';
import GetAvatar from '../lib/GetAvatar.js';

describe('<GetAvatar />', () => {
  it('should render', () => {
    const component = shallow(<GetAvatar width={100} height={100} />);

    expect(component).toMatchSnapshot();
  });
});
