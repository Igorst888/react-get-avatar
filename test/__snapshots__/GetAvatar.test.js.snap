// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GetAvatar /> should render 1`] = `
<div
  style={
    Object {
      "backgroundColor": "black",
      "height": "100px",
      "position": "relative",
      "width": "100px",
    }
  }
>
  <span
    style={
      Object {
        "color": "white",
      }
    }
  >
    Webcam not supported
  </span>
</div>
`;
